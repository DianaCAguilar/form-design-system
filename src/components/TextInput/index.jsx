import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { throttle, debounce } from 'throttle-debounce';
import cc from 'classcat';
import IconInput from './IconInput';
import DecoratedInput from './DecoratedInput';

export const throttleValue = 500;

export const throttleInput = (input, throttleFn, debounceFn) => {
  if (input.length < 5 || input.endsWith(' ')) {
    throttleFn(input);
  } else {
    debounceFn(input);
  }
};

const TextInput = React.forwardRef(
  (
    {
      label,
      labelWidth,
      isLabelBold,
      labelPosition = 'top',
      errorText,
      hasError,
      onThrottledChange,
      showRequired,
      showLabel = true,
      onChange = () => {},
      type = 'text',
      IconLeft,
      IconRight,
      after,
      before,
      'aria-label': ariaLabel,
      ...props
    },
    ref
  ) => {
    // Refer to the following for an explanation on
    // the throttling / debouncing approach:
    // https://www.peterbe.com/plog/how-to-throttle-and-debounce-an-autocomplete-input-in-react
    let onThrottledChangeThrottled;
    let onThrottledChangeDebounced;

    const assignThrottleFunctions = () => {
      if (onThrottledChange) {
        onThrottledChangeThrottled = throttle(throttleValue, onThrottledChange);
        onThrottledChangeDebounced = debounce(throttleValue, onThrottledChange);
      }
    };

    const inputOnChange = (e) => {
      e.persist();
      onChange(e);
      if (onThrottledChange) {
        throttleInput(
          e.target.value,
          onThrottledChangeThrottled,
          onThrottledChangeDebounced
        );
      }
    };

    useEffect(assignThrottleFunctions, [onThrottledChange]);

    const Element = showLabel ? 'label' : 'div';

    return (
      <>
        <Element
          className={cc([
            'fdsTextInput-root',
            `fdsTextInput-root--${labelPosition}Label`,
          ])}
        >
          {showLabel && label && (
            <div
              className={cc([
                {
                  'fontWeight--bold': isLabelBold,
                  'margin--bottom--xs': labelPosition === 'top',
                },
                'fdsTextInput-label',
              ])}
              style={labelWidth ? { width: labelWidth } : undefined}
            >
              {label}
              {showRequired && <span className="color--red">&nbsp;*</span>}
            </div>
          )}
          <DecoratedInput after={after} before={before}>
            <IconInput IconLeft={IconLeft} IconRight={IconRight}>
              <input
                {...props}
                ref={ref}
                aria-label={ariaLabel || (showLabel ? label : undefined)}
                aria-invalid={errorText || hasError ? true : undefined}
                onChange={inputOnChange}
                type={type}
                className={cc([
                  {
                    error: errorText || hasError,
                  },
                  'fdsTextInput',
                ])}
              />
            </IconInput>
          </DecoratedInput>
        </Element>
        {errorText && <div className="color--red type--data">{errorText}</div>}
      </>
    );
  }
);

TextInput.displayName = 'TextInput';

TextInput.propTypes = {
  /** Control aria-label */
  'aria-label': PropTypes.string,
  /** Control whether input has error styling */
  hasError: PropTypes.bool,
  /** Control whether label receives a bold treatment */
  isLabelBold: PropTypes.bool,
  /** Displays as red error text below input */
  errorText: PropTypes.string,

  /** Label used for `label` element */
  label: PropTypes.string.isRequired,

  /** Specify width of label (useful for when label is positioned to left */
  labelWidth: PropTypes.string,

  /**
   * FDS Icon _reference_
   * `IconLeft={StarFilledIcon}`
   */
  IconLeft: PropTypes.func,

  /**
   * FDS Icon _reference_
   * `IconRight={StarFilledIcon}`
   */
  IconRight: PropTypes.func,

  /** Input element `type` attribute */
  type: PropTypes.string,

  /** Controls showing asterisk on label, signifying field is required */
  required: PropTypes.bool,

  /** Standard React onChange event */
  onChange: PropTypes.func,

  /** Custom onChange event that throttles / debounces. */
  onThrottledChange: PropTypes.func,

  /** String to place to the left of the input */
  before: PropTypes.string,

  /** String to place to the right of the input */
  after: PropTypes.string,

  labelPosition: PropTypes.oneOf(['top', 'left']),
  showRequired: PropTypes.bool,
  showLabel: PropTypes.bool,
};

export default TextInput;
