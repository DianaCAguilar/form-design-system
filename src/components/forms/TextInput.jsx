import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { throttle, debounce } from 'throttle-debounce';
import IconInput from 'components/forms/IconInput';
import DecoratedInput from 'components/forms/DecoratedInput';

export const throttleValue = 500;

export const throttleInput = (input, throttleFn, debounceFn) => {
  if (input.length < 5 || input.endsWith(' ')) {
    throttleFn(input);
  } else {
    debounceFn(input);
  }
};

const TextInput = ({
  label,
  errorText,
  onThrottledChange,
  required,
  onChange,
  type,
  IconLeft,
  IconRight,
  after,
  before,
  ...props
}) => {
  // Refer to the following for an explanation on
  // the throttling / debouncing approach:
  // https://www.peterbe.com/plog/how-to-throttle-and-debounce-an-autocomplete-input-in-react
  let onThrottledChangeThrottled;
  let onThrottledChangeDebounced;

  const assignThrottleFunctions = () => {
    if (onThrottledChange) {
      onThrottledChangeThrottled = throttle(throttleValue, onThrottledChange);
      onThrottledChangeDebounced = debounce(throttleValue, onThrottledChange);
    }
  };

  const inputOnChange = (e) => {
    e.persist();
    onChange(e);
    throttleInput(e.target.value, onThrottledChangeThrottled, onThrottledChangeDebounced);
  };

  useEffect(assignThrottleFunctions, [onThrottledChange]);

  const Element = label ? 'label' : 'div';

  return (
    <>
      <Element className="fdsTextInput-root">
        {label && (
          <div className="fdsTextInput-label">
            {label}
            {required && <span className="color--red">&nbsp;*</span>}
          </div>
        )}
        <DecoratedInput after={after} before={before}>
          <IconInput IconLeft={IconLeft} IconRight={IconRight}>
            <input
              {...props}
              required={required}
              onChange={inputOnChange}
              type={type}
              className="fdsTextInput"
            />
          </IconInput>
        </DecoratedInput>
      </Element>
      {errorText && <div className="color--red type--data">{errorText}</div>}
    </>
  );
};

TextInput.defaultProps = {
  type: 'text',
  onChange: () => {},
  onThrottledChange: () => {},
};

TextInput.propTypes = {
  errorText: PropTypes.string,

  /** Label used for `label` element */
  label: PropTypes.string,

  /**
   * FDS Icon _reference_
   * `IconLeft={StarFilledIcon}`
   */
  IconLeft: PropTypes.func,

  /**
   * FDS Icon _reference_
   * `IconRight={StarFilledIcon}`
   */
  IconRight: PropTypes.func,

  /** Input element `type` attribute */
  type: PropTypes.string,

  /** Controls showing asterisk on label, signifying field is required */
  required: PropTypes.bool,

  /** Standard React onChange event */
  onChange: PropTypes.func,

  /** Custom onChange event that throttles / debounces. */
  onThrottledChange: PropTypes.func,

  /** String to place to the left of the input */
  before: PropTypes.string,

  /** String to place to the right of the input */
  after: PropTypes.string,
};

export default TextInput;
