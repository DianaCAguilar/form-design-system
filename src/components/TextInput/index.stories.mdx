import React from 'react';
import CaretDownIcon from 'lib/icons/react/CaretDownIcon';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';
import { action } from "@storybook/addon-actions";
import { useForm } from "react-hook-form";
import InputGroup from 'components/InputGroup';
import State from 'util/storybook-docs/State';

import TextInput from '.';

export const Primary = args => (
  <TextInput {...args} label="First Name"  />
)

export const exampleLayout = () => (
    <>
      <Flex>
        <FlexItem>
          <TextInput label="First Name" showRequired />
        </FlexItem>
        <FlexItem>
          <TextInput label="Last Name" showRequired />
        </FlexItem >
      </Flex>
      <br />
      <Flex>
        <FlexItem>
          <TextInput
            type="email"
            label="Email"
            errorText="Enter a valid email address"
            placeholder="e.g. anand@cbinsights.com"
          />
        </FlexItem>
        <FlexItem>
          <TextInput type="password" label="Password" />
        </FlexItem>
        <FlexItem>
          <TextInput type="number" label="Zipcode" after="90210" />
        </FlexItem>
      </Flex>
    </>
)

export const controlledInput = args => (
    <State
      initialValue="Controlled"
      render={(value, setValue) => (
        <TextInput
          {...args}
          label="Controlled"
          showLabel={false}
          value={value}
          onChange={(e) => setValue(e.target.value)}
        />
      )}
    />
)

export const Icons = args => (
  <TextInput
    {...args}
    label="Icons"
    showLabel={false}
    IconLeft={CaretDownIcon}
    IconRight={CaretDownIcon}
  />
)

export const sideLabel = args => (
    <TextInput
      {...args}
      label="Side Label"
      showLabel={false}
      before="before"
      after="after"
    />
)

export const formManagementLibrary = args => {
      const { register, handleSubmit, errors } = useForm();
      return (<div>
        <form onSubmit={handleSubmit(action('submitted'))}>
          <div className="margin--bottom--half">
          <TextInput
            {...args}
            label="Username"
            name="username"
            type="text"
            errorText={errors.username && errors.username.message}
            ref={register({
              validate: value => value !== "admin" || "Nice try!"
            })}
          />
          </div>
          <TextInput
            {...args}
            label="Email"
            name="email"
            type="text"
            errorText={errors.email && errors.email.message}
            ref={register({
              required: 'Required',
              pattern: {
                value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i,
                message: "invalid email address"
              }
            })}
          />
          <br />
          <button type="submit">Submit</button>
        </form>
      </div>
    )
}

formManagementLibrary.parameters = {
  docs: {
    description: {
      story:
        'You may use this input in any form management library you would like. We recommend `react-hook-form`, which is used in the example below:',
    },
  },
};

export const misc = () => (
    <>
      <InputGroup>
        <TextInput before="$" type="number" value="666" label="Misc" showLabel={false} />
        <select name="unit">
          <option value="1">Million</option>
          <option value="2">Thousand</option>
        </select>
      </InputGroup>
      <br />
      <InputGroup>
        <TextInput IconLeft={CaretDownIcon} type="number" value="666" label="Misc" showLabel={false} />
        <select name="unit">
          <option value="1">Million</option>
          <option value="2">Thousand</option>
        </select>
      </InputGroup>
      <br />
      <InputGroup>
        <TextInput IconRight={CaretDownIcon} type="number" value="666" label="Misc" showLabel={false}/>
        <select name="unit">
          <option value="1">Million</option>
          <option value="2">Thousand</option>
        </select>
      </InputGroup>
      <br />
      <TextInput label="Label on left with width" labelPosition="left" labelWidth="100px" />
    </>
    )

export default {
  component: TextInput,
  title: 'components/TextInput',
  parameters: {
    docs: {
      description: {
        component: 'Standard recommended input component. Can be used for simple no frills inputs, as well as complex inputs complete with accessibile labels, Icons, side labels, errorText, throttled events, and more.'
      },
    },
  },
};


