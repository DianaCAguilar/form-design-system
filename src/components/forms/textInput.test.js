import React from 'react';
import { shallow, mount } from 'enzyme';

import TextInput from './TextInput';

describe('TextInput component', () => {
  it('matches snapshot (default)', () => {
    const wrapper = shallow(<TextInput />);
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (default)', () => {
    const wrapper = shallow(
      <TextInput 
        label="label"
        errorText="errorText"
        IconLeft={() => {}}
        IconRight={() => {}}
        onChange={() => {}}
        throttleOnChange={() => {}}
        before="before"
        after="after"
        type="number"
        required
      />
    );
    expect(wrapper).toMatchSnapshot();
  });


});
