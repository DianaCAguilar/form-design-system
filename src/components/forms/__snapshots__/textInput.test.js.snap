// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput component matches snapshot (default) 1`] = `
<Fragment>
  <div
    className="fdsTextInput-root"
  >
    <DecoratedInput>
      <IconInput>
        <input
          className="fdsTextInput"
          onChange={[Function]}
          type="text"
        />
      </IconInput>
    </DecoratedInput>
  </div>
</Fragment>
`;

exports[`TextInput component matches snapshot (default) 2`] = `
<Fragment>
  <label
    className="fdsTextInput-root"
  >
    <div
      className="fdsTextInput-label"
    >
      label
      <span
        className="color--red"
      >
        Â *
      </span>
    </div>
    <DecoratedInput
      after="after"
      before="before"
    >
      <IconInput
        IconLeft={[Function]}
        IconRight={[Function]}
      >
        <input
          className="fdsTextInput"
          onChange={[Function]}
          type="number"
        />
      </IconInput>
    </DecoratedInput>
  </label>
  <div
    className="color--red type--data"
  >
    errorText
  </div>
</Fragment>
`;
